enhancements/ingress/global-options-enable-hsts.md

```diff
@@ -0,0 +1,411 @@
+---
+title: global-options-enable-hsts 
+authors:
+  - "@cholman"
+
+reviewers:
+  - "@danehans" 
+  - "@frobware"
+  - "@knobunc"
+  - "@Miciah"
+  - "@miheer"
+  - "@rfredette"
+  - "@sgreene570"
+
+approvers:
+  - "@frobware"
+  - "@knobunc"
+
+creation-date: 2021-03-08
+
+last-updated: 2021-03-17
+
+status: implementable
+
+see-also:
+  - https://bugzilla.redhat.com/show_bug.cgi?id=1512759
+  - https://bugzilla.redhat.com/show_bug.cgi?id=1430035
+
+replaces:
+
+superseded-by:
+
+---
+
+# Global Options to Enable HTTP Strict Transport Security (HSTS)
+
+This enhancement provides the capability to enable global automatic HSTS
+as a default for all routes.
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Operational readiness criteria is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+In 3.x and 4.x customers can [provide a per-route annotation to enable HSTS](
+https://docs.openshift.com/container-platform/4.4/networking/routes/route-configuration.html#nw-enabling-hsts_route-configuration).  For
+customers with many routes or regulatory compliance issues, the manual per-route annotation is
+problematic.
+
+This enhancement extends the `IngressController` API to allow cluster administrators to
+enable HSTS globally, without having to add an annotation to each route.
+
+## Motivation
+
+HSTS ([RFC 6797](https://tools.ietf.org/html/rfc6797)) policy enforces the use
+of HTTPS in client requests to the host, without having to use an HTTP redirect.
+HSTS provides user protection and is concerned with minimizing security threats based on network
+traffic eavesdropping and man-in-the-middle attacks.  Using a response header
+called `Strict-Transport-Security`, the HTTP response informs clients that a website
+can only be accessed via HTTPS.  
+
+Administrators who are tasked with route management and/or regulatory compliance
+issues want an option to automatically and globally enable HSTS. They also want
+to separately apply HSTS globally to worker nodes and infrastructure nodes (e.g.
+Prometheus, Grafana, web console hosts), while retaining the ability to opt-out
+of HSTS on a per-route basis.
+
+### Goals
+
+This proposal allows administrators to:
+- Continue to enable HSTS per-route (retain the existing feature functionality)
+- Enable HSTS for all routes globally, based on the cluster's ingress domain
+- Enable HSTS per-router
```

Comment:
Is it "per-router" or "per domain"?  This requirement is very confusing.

Reply:
I interpreted the comments in https://issues.redhat.com/browse/NE-310 to suggest that enabling HSTS on a per-ingress controller level would be sufficient. Am I missing something? It's not clear to be if filtering out subdomains of an ingress controller's domain when assigning HSTS settings is strictly necessary to implement this RFE.

Reply:
I will change this to per domain, and rename the field from inScopeRouters to inScopeDomains.

They originally requested per-node application of HSTS, so that only infra and application nodes could be selected.  I believe we need to keep per-domain to help with that requirement.  If you think this adds too much to the scope of the EP, let me know and I will ask Marc and Mike if we can split off the work into another RFE.

Reply:
I don't think this adds too much scope to this enhancement. I'm just not sure if I completely understand why being able to filter out domains for HSTS settings on an ingress controller is helpful. If a user wants to use different HSTS settings for 2 different sets of routes, they would have to have more than one ingress controller, right (and therefore, the scope settings in this enhancement wouldn't really be useful for them)? The scope fields proposed in this enhancement would give a user the ability to only enable HSTS in any capacity for a strict subset of routes served by a particular ingress controller, right (which i'm assuming is behavior we definitely want to implement)?